File: coding-practice/LeetCode 75/Max Consecutive Ones III/C++/Solution.cpp
# include <iostream>
# include <vector>
using namespace std;

class Solution{
	private:
	
	public:
		int longestOnes(vector<int>& nums, int k){
			int left = 0;
			int max_ones = 0;
			int zeros_count = 0;
			
			for(int right = 0; right < nums.size(); right++){
				if(nums[right] == 0){
					zeros_count += 1;
				}
				
				if(zeros_count > k){
					if(nums[left] == 0){
						zeros_count -= 1;
					}
					left += 1;
				}
				
				max_ones = max(max_ones, (right - left + 1));
			}
			return max_ones;
		};
};

int main(){
	Solution sol;
	int arr[] = {0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1};
	int k = 3;
	vector<int> nums(arr, arr + sizeof(arr) / sizeof(int));
	
	cout << k << " longest ones of {";
	for(int i = 0; i < nums.size(); i++){
		cout << nums[i];
		if(i != nums.size() - 1){
			cout << ", ";
		}
	}
	cout << "} is " << sol.longestOnes(nums, k);
};

File: coding-practice/LeetCode 75/Max Consecutive Ones III/Python/Solution.py
nums = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1]
k = 3

class Solution:
    def __init__(self, nums, k) -> None:
        self.nums = nums
        self.k = k

    def longestOnes(self):
        left = max_ones = zeros_count = 0
        for right in range(len(self.nums)):
            if self.nums[right] == 0:
                zeros_count += 1
            
            if zeros_count > self.k:
                if self.nums[left] == 0:
                    zeros_count -= 1
                left += 1
            max_ones = max(max_ones, (right - left + 1))
        return max_ones
    
obj = Solution(nums, k)
longestOnes = obj.longestOnes()
print(f'{k} longest ones of {nums} is : {longestOnes}')